File: hello-go/src/client/audio_client_application.go
package client

import (
	"fmt"
	"os"
	"time"

	"github.com/feuyeux/hello-mmap/hello-go/src/cli"
	"github.com/feuyeux/hello-mmap/hello-go/src/client/core"
	"github.com/feuyeux/hello-mmap/hello-go/src/client/util"
	"github.com/feuyeux/hello-mmap/hello-go/src/logger"
)

// Run executes the audio client application
func Run() {
	// Parse CLI arguments
	config, err := cli.ParseArgs()
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error parsing arguments: %v\n", err)
		os.Exit(1)
	}

	// Initialize logger
	logger.Init(config.Verbose)

	// Log startup information
	logger.Info("Audio Stream Cache Client - Go Implementation")
	logger.Info(fmt.Sprintf("Server URI: %s", config.Server))
	logger.Info(fmt.Sprintf("Input file: %s", config.Input))
	logger.Info(fmt.Sprintf("Output file: %s", config.Output))

	// Get input file size
	fileSize, err := util.GetFileSize(config.Input)
	if err != nil {
		logger.Error(fmt.Sprintf("Failed to get file size: %v", err))
		os.Exit(1)
	}
	logger.Info(fmt.Sprintf("Input file size: %d bytes", fileSize))

	// Initialize performance monitor
	perf := util.NewPerformanceMonitor(fileSize)

	// Connect to WebSocket server
	logger.Phase("Connecting to Server")
	ws, err := core.Connect(config.Server)
	if err != nil {
		logger.Error(fmt.Sprintf("Failed to connect to server: %v", err))
		os.Exit(1)
	}
	defer ws.Close()
	logger.Info("Successfully connected to server")

	// Upload file
	logger.Phase("Starting Upload")
	perf.StartUpload()
	streamID, err := core.Upload(ws, config.Input, fileSize)
	if err != nil {
		logger.Error(fmt.Sprintf("Upload failed: %v", err))
		os.Exit(1)
	}
	perf.EndUpload()
	logger.Info(fmt.Sprintf("Upload completed successfully with stream ID: %s", streamID))

	// Sleep 2 seconds after upload
	logger.Info("Upload successful, sleeping for 2 seconds...")
	time.Sleep(2 * time.Second)

	// Download file
	logger.Phase("Starting Download")
	perf.StartDownload()
	err = core.Download(ws, streamID, config.Output, fileSize)
	if err != nil {
		logger.Error(fmt.Sprintf("Download failed: %v", err))
		os.Exit(1)
	}
	perf.EndDownload()
	logger.Info("Download completed successfully")

	// Sleep 2 seconds after download
	logger.Info("Download successful, sleeping for 2 seconds...")
	time.Sleep(2 * time.Second)

	// Verify file integrity
	logger.Phase("Verifying File Integrity")
	result, err := util.Verify(config.Input, config.Output)
	if err != nil {
		logger.Error(fmt.Sprintf("Verification error: %v", err))
		os.Exit(1)
	}

	if result.Passed {
		logger.Info("✓ File verification PASSED - Files are identical")
	} else {
		logger.Error("✗ File verification FAILED")
		if result.OriginalSize != result.DownloadedSize {
			logger.Error(fmt.Sprintf("  Reason: File size mismatch (expected %d, got %d)",
				result.OriginalSize, result.DownloadedSize))
		}
		if result.OriginalChecksum != result.DownloadedChecksum {
			logger.Error("  Reason: Checksum mismatch")
		}
		os.Exit(1)
	}

	// Generate performance report
	logger.Phase("Performance Report")
	report := perf.GetReport()
	logger.Info(fmt.Sprintf("Upload Duration: %d ms", report.UploadDurationMs))
	logger.Info(fmt.Sprintf("Upload Throughput: %.2f Mbps", report.UploadThroughputMbps))
	logger.Info(fmt.Sprintf("Download Duration: %d ms", report.DownloadDurationMs))
	logger.Info(fmt.Sprintf("Download Throughput: %.2f Mbps", report.DownloadThroughputMbps))
	logger.Info(fmt.Sprintf("Total Duration: %d ms", report.TotalDurationMs))
	logger.Info(fmt.Sprintf("Average Throughput: %.2f Mbps", report.AverageThroughputMbps))

	// Check performance targets
	if report.UploadThroughputMbps < 100.0 || report.DownloadThroughputMbps < 200.0 {
		logger.Warn("⚠ Performance targets not met (Upload >100 Mbps, Download >200 Mbps)")
	}

	// Disconnect
	logger.Info("Disconnected from server")

	// Log completion
	logger.Phase("Workflow Complete")
	logger.Info(fmt.Sprintf("Successfully uploaded, downloaded, and verified file: %s", config.Input))
}

File: hello-nodejs/src/client/audio-client-application.js
#!/usr/bin/env node

/**
 * Audio Client Application - Main entry point for Node.js audio stream client.
 * Coordinates client operations including upload, download, and verification.
 */

import { parseArgs } from "../cli.js";
import { WebSocketClient } from "./core/WebSocketClient.js";
import { UploadManager } from "./core/UploadManager.js";
import { DownloadManager } from "./core/DownloadManager.js";
import { FileManager } from "./core/FileManager.js";
import { PerformanceMonitor } from "./util/PerformanceMonitor.js";
import { VerificationModule } from "./util/VerificationModule.js";
import { ErrorHandler } from "./util/ErrorHandler.js";
import * as logger from "../logger.js";
import fs from "fs/promises";
import path from "path";

/**
 * Main client application class.
 */
export class AudioClientApplication {
  constructor(config) {
    this.config = config;
    this.fileManager = new FileManager();
    this.errorHandler = new ErrorHandler(config.verbose);
  }

  /**
   * Run the client application.
   */
  async run() {
    try {
      // Initialize logger
      logger.init(this.config.verbose);

      // Log startup information
      logger.info("Audio Stream Cache Client - Node.js Implementation");
      logger.info(`Server URI: ${this.config.server}`);
      logger.info(`Input file: ${this.config.input}`);
      logger.info(`Output file: ${this.config.output}`);

      // Get input file size
      const fileSize = await this.fileManager.getFileSize(this.config.input);
      logger.info(`Input file size: ${fileSize} bytes`);

      // Initialize performance monitor
      const perf = new PerformanceMonitor(fileSize);

      // Ensure output directory exists
      const outputDir = path.dirname(this.config.output);
      await fs.mkdir(outputDir, { recursive: true });

      // Connect to WebSocket server
      logger.phase("Connecting to Server");
      const ws = new WebSocketClient(this.config.server);
      await ws.connect();
      logger.info("Successfully connected to server");

      try {
        // Upload file
        logger.phase("Starting Upload");
        perf.startUpload();

        const uploadManager = new UploadManager(ws);
        let lastProgress = 0;
        const streamId = await uploadManager.upload(
          this.config.input,
          fileSize,
          (bytesSent, total, progress) => {
            if (progress >= lastProgress + 25 && progress <= 100) {
              logger.info(
                `Upload progress: ${bytesSent}/${total} bytes (${progress}%)`,
              );
              lastProgress = progress;
            }
          },
        );

        perf.endUpload();
        logger.info(
          `Upload completed successfully with stream ID: ${streamId}`,
        );

        // Sleep 2 seconds after upload
        logger.info("Upload successful, sleeping for 2 seconds...");
        await new Promise((resolve) => setTimeout(resolve, 2000));

        // Download file
        logger.phase("Starting Download");
        perf.startDownload();

        const downloadManager = new DownloadManager(ws);
        lastProgress = 0;
        await downloadManager.download(
          streamId,
          this.config.output,
          fileSize,
          (bytesReceived, total, progress) => {
            if (progress >= lastProgress + 25 && progress <= 100) {
              logger.info(
                `Download progress: ${bytesReceived}/${total} bytes (${progress}%)`,
              );
              lastProgress = progress;
            }
          },
        );

        perf.endDownload();
        logger.info("Download completed successfully");

        // Sleep 2 seconds after download
        logger.info("Download successful, sleeping for 2 seconds...");
        await new Promise((resolve) => setTimeout(resolve, 2000));

        // Verify file integrity
        logger.phase("Verifying File Integrity");
        const verificationModule = new VerificationModule();
        const result = await verificationModule.verify(
          this.config.input,
          this.config.output,
        );

        if (result.passed) {
          logger.info("✓ File verification PASSED - Files are identical");
        } else {
          logger.error("✗ File verification FAILED");
          if (result.originalSize !== result.downloadedSize) {
            logger.error(
              `  Reason: File size mismatch (expected ${result.originalSize}, got ${result.downloadedSize})`,
            );
          }
          if (result.originalChecksum !== result.downloadedChecksum) {
            logger.error("  Reason: Checksum mismatch");
          }
          process.exit(1);
        }

        // Generate performance report
        logger.phase("Performance Report");
        const report = perf.getReport();
        logger.info(`Upload Duration: ${report.uploadDurationMs} ms`);
        logger.info(`Upload Throughput: ${report.uploadThroughputMbps} Mbps`);
        logger.info(`Download Duration: ${report.downloadDurationMs} ms`);
        logger.info(
          `Download Throughput: ${report.downloadThroughputMbps} Mbps`,
        );
        logger.info(`Total Duration: ${report.totalDurationMs} ms`);
        logger.info(`Average Throughput: ${report.averageThroughputMbps} Mbps`);

        // Check performance targets
        if (
          report.uploadThroughputMbps < 100.0 ||
          report.downloadThroughputMbps < 200.0
        ) {
          logger.warn(
            "⚠ Performance targets not met (Upload >100 Mbps, Download >200 Mbps)",
          );
        }

        // Disconnect
        await ws.close();
        logger.info("Disconnected from server");

        // Log completion
        logger.phase("Workflow Complete");
        logger.info(
          `Successfully uploaded, downloaded, and verified file: ${this.config.input}`,
        );
      } finally {
        await ws.close();
      }
    } catch (error) {
      this.errorHandler.handle(error, "AudioClientApplication");
      process.exit(1);
    }
  }
}

// Run if executed directly
if (import.meta.url === `file://${process.argv[1]}`) {
  const config = parseArgs();
  const app = new AudioClientApplication(config);
  app.run();
}
